pkg/reconciler/common/releases.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+
+	mf "github.com/manifestival/manifestival"
+	"golang.org/x/mod/semver"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	KoEnvKey = "KO_DATA_PATH"
+)
+
+var cache = map[string]mf.Manifest{}
+
+// TargetVersion returns the version of the manifest to be installed
+// per the spec in the component. If spec.version is empty, but
+// status.version is set, that will be returned, since the expectation
+// is that the current installation should be reconciled. If both
+// spec.version and status.version are empty, the latest version known
+// to the operator is returned.
+func TargetVersion(instance v1alpha1.KComponent) string {
+	switch {
+	case instance.GetSpec().GetVersion() != "":
+		return instance.GetSpec().GetVersion()
+	case instance.GetStatus().GetVersion() != "":
+		return instance.GetStatus().GetVersion()
+	default:
+		return latestRelease(instance)
+	}
+}
+
+// TargetManifest returns the manifest for the TargetVersion
+func TargetManifest(ctx context.Context, instance v1alpha1.KComponent, stages ...ReconcileStage) (mf.Manifest, error) {
```

Comment:
Why are the reconcile stages part of this call? Don't they do sideeffects? Looking below it seems like you're using this to apply transforms to the manifests, maybe we should more clearly signify that. We don't want "real" reconcile operations on this function (and on `InstalledManifest` for that matter) I believe. That might purely be a naming thing.

Reply:
I might've gotten a little carried away there, but I don't hate it. It was mainly to clean up the calling code, since I was getting the manifest, appending it to the root, and transforming it, not to mention checking for errors between each. And that was just in `FinalizeKind`, but in *both* the eventing and serving reconcilers. It was a lot of redundant code, and there'll be even more, because...

...we'll need to do the exact same thing in `deleteObsoleteResources` for the installed manifest when we compute the diffs to know what resources to delete.

Plus, we'll likely need to include the `ingress` stage in both places as well, and whatever future stages we create to construct the appropriate manifest per the CR spec. 

It just seemed logical to include optional "post processing" stages as parameters to the fetcher. And while I was at it, I realized I could re-use the logic to clean up the `ReconcileKind` function. It seemed like a win to me.

Reply:
Yeah I think it's a naming thing for me. We shouldn't actively reconcile (aka read + write from the API for example) in these functions, right?

Reply:
Why not? It's the only reason we're fetching the manifest at all. :) 

I'm always open to better names, of course.

Reply:
Well my confusion is that I wouldn't expect sideeffects to happen when I call a function named `TargetManifest` that returns a manifest. I do buy applying transformers to it straight away though. "Reconcile" is just overloaded in my head with a fixup-feedback-loop, which I wouldn't expect this to do.

Reply:
Reverted